packages/field-base/test/text-field-mixin.test.js

```diff
@@ -0,0 +1,201 @@
+import { expect } from '@esm-bundle/chai';
+import sinon from 'sinon';
+import { fixtureSync } from '@vaadin/testing-helpers';
+import { PolymerElement, html } from '@polymer/polymer/polymer-element.js';
+import { TextFieldMixin } from '../src/text-field-mixin.js';
+
+customElements.define(
+  'text-field-mixin-element',
+  class extends TextFieldMixin(PolymerElement) {
+    static get template() {
+      return html`
+        <slot name="label"></slot>
+        <slot name="input"></slot>
+        <button id="clearButton">Clear</button>
+        <div part="error-message">
+          <slot name="error-message"></slot>
+        </div>
+        <slot name="helper"></slot>
+      `;
+    }
+
+    get clearElement() {
+      return this.$.clearButton;
+    }
+  }
+);
+
+describe('text-field-mixin', () => {
+  let element, input;
+
+  beforeEach(() => {
+    element = fixtureSync('<text-field-mixin-element></text-field-mixin-element>');
+    input = element.querySelector('[slot=input]');
+  });
+
+  describe('properties', () => {
+    it('should propagate pattern property to the input', () => {
+      element.pattern = '[-+\\d]';
+      expect(input.pattern).to.equal('[-+\\d]');
+    });
+
+    it('should propagate maxlength property to the input', () => {
+      element.maxlength = 8;
+      expect(input.maxLength).to.equal(8);
+    });
+
+    it('should propagate minlength property to the input', () => {
+      element.minlength = 8;
+      expect(input.minLength).to.equal(8);
+    });
+  });
+
+  describe('validation', () => {
+    it('should not change invalid property if no constraints are set', () => {
+      element.validate();
+      expect(element.invalid).to.be.false;
+      element.invalid = true;
+      element.validate();
+      expect(element.invalid).to.be.true;
+    });
+
+    it('should validate the field when invalid and required is set', () => {
+      element.invalid = true;
+      const spy = sinon.spy(element, 'validate');
+      element.required = true;
+      expect(spy.calledOnce).to.be.true;
+    });
+
+    it('should validate the field when invalid after minlength is changed', () => {
```

Comment:
This tests that validate is called when a constraint changes and the input is invalid.

However it seems [this line](https://github.com/vaadin/web-components/pull/2245/files#diff-c31ae29deeca08e5b4346ac104ba5e91d792f09e819e03042ee90856012efa11R76) not tested: We probably also want to check that validate is not called when any of the constraints change, and the input is valid.

Reply:
Removed the related code, as mentioned in the comment above.

Reply:
Added the missing test to ensure `invalid` is not set when we set `required` to true on the valid input.